### PR TITLE
Added sampling to dnscap

### PR DESCRIPTION
**Pull request by chegger**
Added sampling to dnscap. With the new -q option, the user is now able to decide whether the output should be sampled as follows: Only every nth query is output and only responses that correspond to one of these sampled queries will be output.
e.g.:
dnscap -r ./file.pcap -g -q 10 will output only every 10th query and these queries' corresponding responses.
Due to the nature of this sampling, the -q option is not compatible with -s r (of which the user would informed by usage()).

Updated dnscap.1.in to match this addition.

Added uthash/uthash.h since I needed a hashtable for the sampling to work. The dnscap and uthash licenses should be compatible.

---

I've made a very similar pull request at an outdated repository about 16 days ago already:
https://github.com/DNS-OARC/dnscap/pull/10

Now that I' have based my work on the DNS-OARC repo, there are still two issues to talk about:
- "Making the -q option more versatile"
  According to an unfinished master's thesis, sampling the way I implemented, is as suitable as a random distribution for DNS traffic. Since this thesis is not published yet, I cannot provide a source to it.
  Also, might it not be difficult to produce good random numbers at such high package processing speeds? A way could be using the content of the single packets to produce random numbers from (like the transaction id)?.
  All in all, I'm worried that certain packages could be favored over others.
- "Using the already existing generic hashtable  dnscap/plugins/rssm/hashtable.c"
  Simply put, I think it would mean a lot of unnecessary work that I probably can't do properly, since I don't have a lot of experience yet. This would probably make the hash table rather inefficient.
  I have absolutely no objection to somebody else doing so, though (although uthash seems quite OK to me).

### CODE REVIEW

**Comment by jelu** [comment]
You are calling this a module (or plugin) but it is not one, is there a reason you did not implement this under `plugins/sampling/` ?

File: src/dnscap.1.in [hunk 1]
```diff
@@ -53,6 +53,7 @@
 .Op Fl E Ar datetime
 .Op Fl P Ar plugin.so
 .Op Fl U Ar str
+.Op Fl q Ar unsigned int
```
Comment (jelu) [inline src/dnscap.1.in]: The phrasing `unsigned int` is not very user friendly, use something like `nth` instead.

File: src/dnscap.1.in [hunk 2]
```diff
@@ -353,6 +354,11 @@ Enable immediate mode on interfaces.
 Append "and
 .Ar str
 " to the pcap filter.
+.It Fl q Ar unsigned int
```
Comment (jelu) [inline src/dnscap.1.in]: Same here.

File: src/dnscap.1.in [hunk 3]
```diff
@@ -353,6 +354,11 @@ Enable immediate mode on interfaces.
...
+Causes the output to be sampled after the application of all other filters.
+Only every nth dns initiation will be output and responses are only output
+if they correspond to one of the sampled queries. This option cannot be
+used with option.
```
Comment (jelu) [inline src/dnscap.1.in]: Last sentence is unclear, specify the options that can't be used.

File: src/dnscap.c [hunk 4]
```diff
@@ -1099,6 +1130,11 @@ parse_args(int argc, char *argv[]) {
 		usage("the -L and -l options are mutually exclusive");
 	if (background && (dumptrace || preso))
 		usage("the -b option is incompatible with -d and -g");
+        if (sample && chooseSidesResponse)
+        {
+                if (!((dir_wanted & DIR_INITIATE) != 0) && ((dir_wanted & DIR_RESPONSE) != 0))
```
Comment (jelu) [inline src/dnscap.c]: Why check this? If both are true then both options have been used.

File: src/dnscap.c [hunk 5]
```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
 			abort();
 		}
 	}
-	output(descr, from, to, proto, flags, sport, dport, ts,
-	    pkt_copy, olen, dnspkt, dnslen);
+/*Sample Module*/
+        if (sample == TRUE)
+        {
+                ns_msg dnsmsgSample;
+                ns_initparse(dnspkt,dnslen,&dnsmsgSample);
```
Comment (jelu) [inline src/dnscap.c]: This is part of libbind and dnscap can be compiled without libbind.

File: src/dnscap.c [hunk 6]
```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
...
+                samplePacket *currentQuery;
+                void *ipAddrBufferFrom = &from.u;
+                void *ipAddrBufferTo = &to.u;
```
Comment (jelu) [inline src/dnscap.c]: There is no need for these `void *` you can type cast it in one go below.

File: src/dnscap.c [hunk 7]
```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
...
+                keylen = offsetof(samplePacket,to)      //keylen is used to define which fields of the hash structure are added
+                        + sizeof(*toBuffer)                    //as a compound key. Here, the key is composed of all fields between (and including)
+                        - offsetof(samplePacket,from);  //samplePacket->to and samplePacket->from (from, sport, dport, transaction_id, to)
+
+                if(dns.qr == 0)
+                {
+                        querycount++;
+                        if(querycount % sampleAmount == 0)
+                                {
+                                        currentQuery = malloc(sizeof(*currentQuery));
```
Comment (jelu) [inline src/dnscap.c]: You do not check if `malloc()` was successful.

File: src/dnscap.c [hunk 8]
```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
...
+                                        currentQuery->dport = dport;
+                                        currentQuery->transaction_id = ns_msg_id(dnsmsgSample);
+
+                                        HASH_ADD(hh,allSampleQueries,from,keylen,currentQuery);
+                                        output(descr,from,to,proto,flags,sport,dport,ts,pkt_copy,olen,dnspkt,dnslen);
+                                }
+                }
+                else
+                {
+                        sample_lookup_key *lookup_key = (sample_lookup_key*)malloc(sizeof(*lookup_key));;
```
Comment (jelu) [inline src/dnscap.c]: You do not check if `malloc()` was successful.

File: src/dnscap.c [hunk 9]
```diff
@@ -2339,8 +2375,61 @@ network_pkt(const char *descr, my_bpftimeval ts, unsigned pf,
...
+                        - offsetof(samplePacket,from);  //samplePacket->to and samplePacket->from (from, sport, dport, transaction_id, to)
+
+                if(dns.qr == 0)
+                {
+                        querycount++;
+                        if(querycount % sampleAmount == 0)
+                                {
+                                        currentQuery = malloc(sizeof(*currentQuery));
+                                        memset(currentQuery, 0, sizeof(*currentQuery));
+                                        currentQuery->from = *fromBuffer;
```
Comment (jelu) [inline src/dnscap.c]: What are meant to be in `samplePacket->from` ?
From the struct definition `from` is just an `unsigned` and this would essentially put the first `unsigned` from the `fromBuffer` into `currentQuery`. That would be the first part of `iaddr.u` and that is only the IPv4 address (or a very small part of it) which means this does not work for IPv6.

**Review by jelu — CHANGES_REQUESTED** [review]
Overall, please use more spaces in both control sections and argument lists and try to mimic the coding style of the rest of the code.

File: src/dnscap.c (hunk 4 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: It should be incompatible with -s r. I fixed this flaw.
I'm actually sorry, I thought I had tested this..

File: src/dnscap.c (hunk 6 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: Oh okay, I initially tried to cast it in one go, but i got a compiler error. Ill try again, thank you!

File: src/dnscap.c (hunk 7 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: Mhm okay, thank you! I'll add all malloc checks that need to be added.

File: src/dnscap.c (hunk 8 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: Mhm okay, thank you! I'll add all malloc checks that need to be added.

File: src/dnscap.1.in (hunk 2 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.1.in]: I see, I'll change it!

File: src/dnscap.1.in (hunk 3 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.1.in]: I'm very sorry, this must have been a copy and paste error. The last sentence is only half of the sentence it should be, I'll correct that.

File: src/dnscap.c (hunk 9 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: `samplePacket->from` should contain IP addresses (both IPv4 and v6 would be optimal).

On my system `unsigned` is 32 bits (which is the size of IPv4 adresses), so it should be able to store a full IPv4. I'll change it to `uint_32` so it also works on all systems.
As to IPv6, I did not find out how to handle such big numbers properly. I thought it would be better to take a small portion of the IPv6 addresses than not to work with IPv6 at all.

Comment (jelu) [inline src/dnscap.c]: Sorry you need to find a way to support both v4 and v6 otherwise I will reject the pull request. I do not know how exactly uthash works but maybe you can use the either `iaddr` which `from` is.

File: src/dnscap.c (hunk 8 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: I see, i should also you calloc instead of malloc, since I initialize the pointers with 0s afterwards anyway.

**Comment by SuperAmbitiousItGuy** [comment]
> You are calling this a module (or plugin) but it is not one, is there a reason you did not implement this under `plugins/sampling/` ?

You are right, this is not a module or a plugin, that's why it's not implemented in `plugins/sampling/`. The reason why I call it module is that I lack better terminology. What should it be called? An Addition?

**Comment by jelu** [comment]
Just "sampling" is sufficient.

File: src/dnscap.c (hunk 8 above)
Comment (jelu) [inline src/dnscap.c]: Yes, one `calloc()` instead of `malloc()` and `memset()`.

**Comment by jelu** [comment]
I also see that you have created this pull request from your `develop` branch, please in the future do pull requests from feature branches.
```
git checkout -b feature_name develop
git commit ...
git push -u origin feature_name
```

**Comment by SuperAmbitiousItGuy** [comment]
> I also see that you have created this pull request from your develop branch, please in the future do pull requests from feature branches.

Okay, thanks for letting me know! As of now, I haven't even come to set up github locally on my computer yet. I'm still exclusively using the web interface. Takes some time to get into it.

File: src/dnscap.c (hunk 9 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: I initially used `from` but I came to the conclusion that it doesn't work properly. I'll give it another try though. If I don't get it to work like this, I'll look for another way to support IPv6 addresses.

In case I can't get it to work with IPv6 at all, I'll inform you, so you can reject this pull request.

File: src/dnscap.c (hunk 5 above)
Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: You mean `ns_msg` and `ns_initparse`? Is there another way to retrieve the transaction id of a query? I've seen `ns_initparse` being used in the output functions in `dump_dns.c`, so I assumed it is safe to use.

Comment (jelu) [inline src/dnscap.c]: Look at the top `#if HAVE_NS_INITPARSE && ...`, you will need to disable all the sampling code if libbind is missing or create your own functions for parsing the required information.

Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: For now I'll disable the sampling code if libbind is missing. In future i might try to write my own parsing functions

File: src/dnscap.c [hunk 10]
```diff
@@ -1132,8 +1132,8 @@ parse_args(int argc, char *argv[]) {
 		usage("the -b option is incompatible with -d and -g");
         if (sample && chooseSidesResponse)
         {
-                if (!((dir_wanted & DIR_INITIATE) != 0) && ((dir_wanted & DIR_RESPONSE) != 0))
-                        usage("the -q option is incompatible with -s r");
+		if(((dir_wanted & DIR_RESPONSE) != 0) && ((dir_wanted & DIR_INITIATE) == 0))
+			usage("the -q option is incompatible with -s r");
```
Comment (jelu) [inline src/dnscap.c]: On second glance I do not see the reason to have `chooseSidesResponse`, `dir_wanted` is set to both direction at start and you really only need to check if `DIR_RESPONSE` is not set.

File: src/dnscap.c (hunk 9 above)
Comment (wessels) [inline src/dnscap.c]: Look at plugins/rssm/rssm.c for an example of how to make a hash table with both v4 and v6 addresses ("my_hashtbl").

dnscap_common.h defines 'iaddr' which you should use to store v4/v6 addrs.  

rssm.c (and other plugin .c files) have iaddr_hash().

At some point it would make sense to move hashtbl.c and supporting functions out of plugins and into src.

Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: Found a way that will work perfectly fine. Ill commit it on Monday!

Comment (jelu) [inline src/dnscap.c]: Sorry but you have implemented this in a very dangerous way and the use of `AMOUNT_32BIT_IN_128BIT` is unneeded. You are copying in and out of different structures without any check of structure sizes and seeing that the hash can take binary blobs as keys you should be able to put two `iaddr` structures within the hash entry for both to and from v4/v6 addresses.

File: src/dnscap.c (hunk 5 above)
Comment (jelu) [inline src/dnscap.c]: The sampling code is still compiled even if libbind does not exist, checking if `sample` is true or not does not remove the dependency of the functions.

Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: Yes but in `parse_args()`  `sample` is only set to true if libbind does exist, by checking `#if HAVE_NS_INITPARSE && HAVE_NS_PARSERR && HAVE_NS_SPRINTRR` . Is this checkback not suitable?

Comment (jelu) [inline src/dnscap.c]: You do know the difference between `if()` and `#if` right?

Comment (SuperAmbitiousItGuy) [inline src/dnscap.c]: I do and I just realized where my mistake in thinking was at. I will work on it right away.

**Comment by SuperAmbitiousItGuy** [comment]
Well, @jelu what do you think now?

**Comment by jelu** [comment]
Sorry @SuperAmbitiousItGuy but I am pressed for time on other issues this week, hope to look at this next week and I won't be including this in v1.1.0 so it will not be merged just yet even if approved.
